File: algorithms/Binary Search/binary_search.ex
defmodule BinarySearch do
  def search(list, item) do
    low = 0
    high = length(list) - 1

    do_search(list, item, low, high)
  end

  defp do_search(_, _, low, high) when high < low, do: nil

  defp do_search(list, item, low, high) do
    mid = div(low + high, 2)
    guess = Enum.at(list, mid)

    cond do
      guess == item -> mid
      guess > item -> do_search(list, item, low, mid - 1)
      true -> do_search(list, item, mid + 1, high)
    end
  end
end

my_list = [1, 3, 5, 7, 9]

IO.puts(BinarySearch.search(my_list, 3))
# => 1

IO.inspect(BinarySearch.search(my_list, -1))
# => nil

File: algorithms/Binary Search/binary_search.cpp
#include <iostream> 
using namespace std;

int binarySearch(int a[], int l, int r, int x);

int main()
{
    int n;
    cin >> n;
    int a[n];
    for(int i = 0; i < n; i++)
        cin >> a[i];
    int x;
    cin >> x;
    int result = binarySearch(a, 0, n - 1, x);
    if(result == -1)
        cout << "Not found";
    else
        cout << "Found";
    return 0;
}
int binarySearch(int a[], int l, int r, int x) 
{ 
    if (r >= l) { 
        int mid = l + (r - l) / 2; 
  
        
        if (a[mid] == x) 
            return mid; 
  
        
        if (a[mid] > x) 
            return binarySearch(a, l, mid - 1, x); 
   
        return binarySearch(a, mid + 1, r, x); 
    }   
    return -1; 
} 